### ISSUE TITLE
Provide a list of packages for developer to install? 

### ISSUE THREAD

**Issue by C-Kenny** [processing/processing#5937]
How to specify a list of dependencies that have to be installed before developing a sketch? I understand that if you export the sketch for others it will include them but is there a Processing equivalent of Python's pip package manager?

Right now I'm thinking of adding a list of dependencies that they can add through the GUI library installed but it doesn't feel quite right.

**Comment by benfry** [comment]
I don't understand… what dependencies? The point of the PDE is that people can begin coding immediately, without additional installation required.

**Comment by jeremydouglass** [comment]
I think I understand. For example, if a sketch includes:

```
import controlP5.*;
```

Then you want to be able to send that sketch -- not an exported app -- to another computer that does not have the ControlP5 contributed library already installed, then the sketch will list the missing dependencies in the Errors tab. Is that right? You want to know if there is an automatic way to communicate the installation requirements other than exporting, or bundling them with the sketch, or having the user resolve each error in the errors tab by clicking to install or by installing through Contributions manager.

Note that you can also include full copies of library jars in the sketch /code folder if you want to -- in particular, if these are libraries that aren't available through Contributions manager.

This sounds like it is kind of a feature request...?, but might first be good to have as a discussion on the https://discourse.processing.org forum.

**Comment by C-Kenny** [comment]
Yeah @jeremydouglass is on point, if the sketch includes a library and a new person/environment tries to run the sketch it won't work, as there are libraries missing.

In other programming environments, you can specify libraries/dependencies in an external file and a package manager will install the required libraries for you. I was hoping Processing had similar functionality.

Node example:
`$ npm install` which pulls dependencies from package.json

Python example:
`$ pip install -r requirements.txt`

@benfry Any sketch that imports external libraries isn't easy to share with another developer, as they'll have to go through the Contributions Manager and figure out what packages they are missing.

I was hoping Processing would have something like:

`$ processing --install-required-libraries required_libraries.txt` and then you'd be able to share the source code of your sketch with other developers, who can install the external dependencies with that command.

Hope that makes sense. Maybe the Processing forum would be a better place.

**Comment by kfrajer** [comment]
If you implement something like this, you will also need to create a lock package concept where the 
libraries/dependencies are locked to specific versions. Does Processing support library versioning?

From my experience, Processing is not to advance as to have a file descriptor of dependencies. I could think in exploring maven/graddle concepts but I am probably asking too much (You are coming from the Python and Node.js but you should consider looking at Java implementations)

This could be a possible topic for GSoC '20?

Kf
